File: sdk/python/stagehand/_providers.py
from __future__ import annotations

import os
from typing import Final

ENV_OPENAI_HOSTS: Final[str] = "STAGEHAND_OPENAI_HOSTS"
_DEFAULT_OPENAI_HOSTS: Final[frozenset[str]] = frozenset({"api.openai.com"})


def configured_openai_hosts() -> frozenset[str]:
    raw = os.environ.get(ENV_OPENAI_HOSTS, "")
    configured = {host.strip().lower() for host in raw.split(",") if host.strip()}
    return frozenset(_DEFAULT_OPENAI_HOSTS | configured)


def is_openai_host(hostname: str | None) -> bool:
    if hostname is None:
        return False

    return hostname.strip().lower() in configured_openai_hosts()

File: sdk/typescript/src/providers.ts
export const ENV_OPENAI_HOSTS = 'STAGEHAND_OPENAI_HOSTS';

const DEFAULT_OPENAI_HOSTS = new Set(['api.openai.com']);

export function configuredOpenAIHosts(env: NodeJS.ProcessEnv = process.env): Set<string> {
  const configured = new Set<string>();
  const raw = env[ENV_OPENAI_HOSTS] ?? '';
  for (const host of raw.split(',')) {
    const normalized = host.trim().toLowerCase();
    if (normalized.length > 0) {
      configured.add(normalized);
    }
  }

  return new Set([...DEFAULT_OPENAI_HOSTS, ...configured]);
}

export function isOpenAIHost(
  hostname: string | undefined,
  env: NodeJS.ProcessEnv = process.env
): boolean {
  if (hostname === undefined) {
    return false;
  }

  return configuredOpenAIHosts(env).has(hostname.trim().toLowerCase());
}

export function openAIOperationFromURL(
  method: string,
  url: URL | undefined,
  env: NodeJS.ProcessEnv = process.env
): string | undefined {
  if (url === undefined || !isOpenAIHost(url.hostname, env) || method !== 'POST') {
    return undefined;
  }

  if (url.pathname === '/v1/chat/completions') {
    return 'chat.completions.create';
  }

  if (url.pathname === '/v1/responses') {
    return 'responses.create';
  }

  return undefined;
}
